File: Genericidad/Exercise5/Visual Studio Code-Python/Pila.py
from typing import Generic, TypeVar, List
T = TypeVar('T')

class Pila(Generic[T]):
    def __init__(self):
        self.L: List[T] = []  
    def push(self, g: T) -> None:
        self.L.append(g)
    def pop(self) -> T:
        if not self.L:
            raise IndexError("La pila está vacía")
        return self.L.pop()
    def mostrar(self) -> None:
        print("Pila Prueba: ")
        print("Elementos de la pila:", self.L)
    def is_empty(self) -> bool:
        return len(self.L) == 0
    def size(self) -> int:
        return len(self.L)
    def peek(self) -> T:
        if not self.L:
            raise IndexError("La pila está vacía")
        return self.L[-1]

pilant = Pila[int]()
pilant.push(10)
pilant.push(20)
pilant.push(30)
pilant.mostrar() 
print("Desapilando:", pilant.pop()) 
pilant.mostrar() 
pilacad = Pila[str]()
pilacad.push("Gallina")
pilacad.push("Turuleca")
pilacad.mostrar()
print("Desapilando:", pilacad.pop()) 
pilacad.mostrar()

class Persona:
        def __init__(self, nombre: str, edad: int):
            self.nombre = nombre
            self.edad = edad

        def __repr__(self):
            return f"Persona(nombre='{self.nombre}', edad={self.edad})"

pilaper = Pila[Persona]()
pilaper.push(Persona("Evelyn", 35))
pilaper.push(Persona("Roberto",30))
pilaper.mostrar() 
print("Desapilando:", pilaper.pop()) 
pilaper.mostrar()

File: Genericidad/Exercise5/Eclipse-Java/src/genericidad/Pila.java
package genericidad;
import java.util.ArrayList;
import java.util.List;

public class Pila<T> {
    private List<T> Z; 

    public Pila() {
        this.Z = new ArrayList<>();
    }

    public void push(T x) {
        Z.add(x);
    }

    public T pop() {
        if (Z.isEmpty()) {
            throw new IllegalStateException("La pila está vacía");
        }
        return Z.remove(Z.size() - 1);
    }

    public void mostrar() {
    	System.out.println("Pila Prueba: ");
        System.out.println("Elementos de la pila: " + Z);
    }

    public boolean isEmpty() {
        return Z.isEmpty();
    }

    public int size() {
        return Z.size();
    }

    public T peek() {
        if (Z.isEmpty()) {
            throw new IllegalStateException("La pila está vacía");
        }
        return Z.get(Z.size() - 1);
    }

    public static void main(String[] args) {
        Pila<Integer> pila1 = new Pila<>();
        pila1.push(15);
        pila1.push(30);
        pila1.push(45);
        pila1.mostrar();

        System.out.println("Desapilando: " + pila1.pop());
        pila1.mostrar();

        Pila<String> pila2 = new Pila<>();
        pila2.push("Tralalero");
        pila2.push("Tralala");
        pila2.mostrar();

        System.out.println("Desapilando: " + pila2.pop());
        pila2.mostrar();

        class Persona {
            String nombre;
            int edad;

            public Persona(String nombre, int edad) {
                this.nombre = nombre;
                this.edad = edad;
            }

            @Override
            public String toString() {
                return "Persona{" + "Nombre : '" + nombre + "', Edad :" + edad + '}';
            }
        }

        Pila<Persona> pila3 = new Pila<>();
        pila3.push(new Persona("Daniela",30));
        pila3.push(new Persona("Hugo", 50));
        pila3.mostrar();

        System.out.println("Desapilando: " + pila3.pop());
        pila3.mostrar();
    }
}
